pertpy/plot/_mixscape_violin.py

```diff
@@ -0,0 +1,204 @@
+from typing import Literal, Sequence
+
+import numpy as np
+import pandas as pd
+from anndata import AnnData
+from matplotlib import pyplot as pl
+from matplotlib.axes import Axes
+from scanpy import get
+from scanpy._settings import settings
+from scanpy._utils import _check_use_raw, sanitize_anndata
+from scanpy.plotting import _utils
+from scanpy.plotting._utils import setup_axes
+
+
+def mixscape_violin(
+    adata: AnnData,
+    target_gene_idents: str | list[str],
+    keys: str | Sequence[str] = "mixscape_class_p_ko",
+    groupby: str | None = "mixscape_class",
+    log: bool = False,
+    use_raw: bool | None = None,
+    stripplot: bool = True,
+    hue: str | None = None,
+    jitter: float | bool = True,
+    size: int = 1,
+    layer: str | None = None,
+    scale: Literal["area", "count", "width"] = "width",
+    order: Sequence[str] | None = None,
+    multi_panel: bool | None = None,
+    xlabel: str = "",
+    ylabel: str | Sequence[str] | None = None,
+    rotation: float | None = None,
+    show: bool | None = None,
+    save: bool | str | None = None,
+    ax: Axes | None = None,
+    **kwds,
+):
+    """Violin plot using mixscape results. Need to run `pt.tl.mixscape` first.
+
+    Args:
+        adata: The annotated data object.
+        target_gene: Target gene name to plot.
+        keys: Keys for accessing variables of `.var_names` or fields of `.obs`. Default is 'mixscape_class_p_ko'.
+        groupby: The key of the observation grouping to consider. Default is 'mixscape_class'.
+        order: Order in which to show the categories.
+        xlabel: Label of the x axis. Defaults to `groupby` if `rotation` is `None`, otherwise, no label is shown.
+        ylabel: Label of the y axis. If `None` and `groupby` is `None`, defaults to `'value'`. If `None` and `groubpy` is not `None`, defaults to `keys`.
+        stripplot: Add a stripplot on top of the violin plot.
+        log: Plot on logarithmic axis.
+        use_raw: Whether to use `raw` attribute of `adata`. Defaults to `True` if `.raw` is present.
+        show: Show the plot, do not return axis.
+        save: If `True` or a `str`, save the figure. A string is appended to the default filename. Infer the filetype if ending on {`'.pdf'`, `'.png'`, `'.svg'`}.
+        ax: A matplotlib axes object. Only works if plotting a single component.
+        **kwds: Additional arguments to `seaborn.violinplot`.
+
+    Returns:
+        A :class:`~matplotlib.axes.Axes` object if `ax` is `None` else `None`.
+    """
+
+    if isinstance(target_gene_idents, str):
+        mixscape_class_mask = adata.obs[groupby] == target_gene_idents
+    elif isinstance(target_gene_idents, list):
+        mixscape_class_mask = np.full_like(adata.obs[groupby], False, dtype=bool)
+        for ident in target_gene_idents:
+            mixscape_class_mask |= adata.obs[groupby] == ident
+    adata = adata[mixscape_class_mask]
+
+    import seaborn as sns  # Slow import, only import if called
```

Comment:
I am surprised that the seaborn import is slow? Think it's mostly just wrappers of a few matplotlib functions?

Reply:
No idea, I copied this from scanpy.pl.violin directly.